### ISSUE TITLE
Needs response label based constraint

### ISSUE THREAD

**Issue by prashanthellina** [deep-compute/gitkanban#136]
We are able to ensure now that all work progresses towards closure with the aid of Gitkanban alerts (need to finish supporting Inbox and Analysis alerts) and have to start focusing on the speed of delivery aspect. Towards that, I notice that issues in "next" queue are slowing down way too much especially when they are blocking on simple communication. eg: A person is asked a question which will take 5 mins to answer but that comment satisfies Gitkanban for 3 more days. Only then is the next alert sent out. There must be a way to force request an alert within a couple of hours even if the issue is in the next queue. I am thinking we can add something like "needs-update-in-1h", "needs-update-in-2h", ""needs-update-by-eod" kind of labels. These can be applied to ANY issue in ANY queue. From the time of such a label being assigned, if no update happens on that issue within the specified duration, we can raise an alert. 

- [x] Decide list of labels required (unlike moveto, we don't need long durations - max can be 1d)
- [x] Estimate API usage
- [x] Implement and apply constraint (should work across queues just like moveto labels).

**Comment by Rajendra-R** [comment]
* We will work on this issue to day.

**Comment by Rajendra-R** [comment]
* Started working on this issue.

**Comment by Rajendra-R** [comment]
@prashanthellina I would like to create the labels which are mentioned given below, could you please check and let me know if any changes required.
```
needs-update-in-1h,
needs-update-in-2h,
needs-update-in-4h,
needs-update-in-6h,
needs-update-by-eod
```
* To Implement this do we create any new `sub-command` or `new_constraint` for this?
* Do we need to specify these labels separately in our config file like snooze_labels?

**Comment by prashanthellina** [comment]
Looks good @Rajendra-R . Let us do new sub-command for this. No constraint is required. Is there a special constrain for moveto labels?

**Comment by Rajendra-R** [comment]
No, we didn't have any constraint for `moveto-inprogress-*` label.

**Comment by Rajendra-R** [comment]
@prashanthellina If we want to raise an alert for `needs-update-in-*` as you told we didn't require any constraint and you said as new sub-command.
* How the alert title looks for this alert?
* What kind of labels we need to assign for that alert?
* Do we require escalation for this alerts?
* Where do we specify all these configs?

**Comment by prashanthellina** [comment]
@Rajendra-R Good questions. Please think about these and propose the design.

**Comment by Rajendra-R** [comment]
* Started working on this issue.

**Comment by Rajendra-R** [comment]
* As per discussion with @prashanthellina we designed a config format for needs-update labels. Could you please check and let you know.
```
needs_update: {
    "labels": [
         {
              "name": "needs-update-in-1h",
              "time": "1h"
         },
         {
              "name": "needs-update-in-2h",
              "time": "2h"
         },
         {
              "name": "needs-update-in-4h",
              "time": "4h"
         },
         {
              "name": "needs-update-in-6h",
              "time": "6h"
         },
         {
              "name": "needs-update-in-eod",
              "time": "eod"
         }
     ],
     "constraint": {
         "name": "{labels.name}",
         "queue": null,
         "time_since_activity": "{labels.time}",
         "priority": "low",
         "frequency": "hourly",
         "continue": false,
         "title": "No comment for {issue_title} ({issue_id}) since {time_since_activity}",
          "description": null,
          "follow_ups": [
              {
                   "time_elapsed": "1h",
                   "priority": "medium",
                   "message": "Please post an update"
              },
              {
                   "time_elapsed": "1h",
                   "priority": "high",
                   "message": "Please post an update to prevent escalation"
              },
              {
                   "time_elapsed": "1h",
                   "priority": "medium",
                   "message": "Please post an update to prevent further escalation",
                   "escalate": true
              },
          ]
     }
}
```

**Comment by prashanthellina** [comment]
@Rajendra-R  Looks good

"time_since_activity": "{labels.time}" --> Can you explain more about how labels.time is computed?

**Comment by Rajendra-R** [comment]
From the above labels while iterating which label is matched for an issue will get that label `time` and `name` into the constraint.

**Comment by prashanthellina** [comment]
@Rajendra-R Got it. Let us use `label_name` and `label_time` so it is inline with `issue_title`, `issue_id` etc.

**Comment by Rajendra-R** [comment]
* We are working on bug issues once it is done will work on this issue.

**Comment by Rajendra-R** [comment]
We will work this issue in this week.

**Comment by Rajendra-R** [comment]
* Started working on this issue.

**Comment by Rajendra-R** [comment]
* Logic Implementation is done.
* Testing is in-progress.

**Comment by Rajendra-R** [comment]
* As per discussion with Prashanth Instead of sub-command we can make this feature under `check_constraints` itself.

**Comment by Rajendra-R** [comment]
* Moved to another issue because of high prior bugs now revert back to this issue.

**Comment by Rajendra-R** [comment]
* Logic implementation is done.
* Testing is in progress.

**Comment by Rajendra-R** [comment]
* Testing is in-progress.

**Comment by Rajendra-R** [comment]
* Testing is done giving PR

**Comment by Rajendra-R** [comment]
* While Implementing this feature we added a new key `label` to the constraint because to request a repo with a label and used a new key called `time_since_labeled`.
* The final config setting for `needs_update` as given below.
```
  "needs_update": {
    "labels": [
      {
        "name": "needs-update-in-1h",
        "time": "1h"
      },
      {
        "name": "needs-update-in-2h",
        "time": "2h"
      },
      {
        "name": "needs-update-in-4h",
        "time": "4h"
      },
      {
        "name": "needs-update-in-6h",
        "time": "6h"
      },
      {
        "name": "needs-update-in-eod",
        "time": "eod"
      }
    ],
    "constraint": {
      "name": "{label_name}",
      "queue": null,
      "label": "{label_name}",
      "time_since_labeled": "{label_time}",
      "priority": "low",
      "frequency": "hourly",
      "continue": false,
      "title": "No comment for {issue_title} ({issue_id}) since {time_since_labeled}",
      "description": null,
      "follow_ups": [
        {
          "time_elapsed": "1h",
          "priority": "medium",
          "message": "Please post an update"
        },
        {
          "time_elapsed": "1h",
          "priority": "high",
          "message": "Please post an update to prevent escalation"
        },
        {
          "time_elapsed": "1h",
          "priority": "medium",
          "message": "Please post an update to prevent further escalation",
          "escalate": true
        }
      ]
    }
  }
```

**Comment by Rajendra-R** [comment]
* This issue is resolved so we can close this ticket.
